lib/collection/src/collection/shard_transfer.rs

```diff
@@ -202,72 +198,56 @@ impl Collection {
-        {
-            let shard_promoted =
-                transfer::driver::finalize_partial_shard(shard_holder, transfer.shard_id).await?;
-            log::debug!(
-                "shard_promoted: {shard_promoted}, shard_id: {}, peer_id: {}",
-                transfer.shard_id,
-                self.this_peer_id,
-            );
-        }
+        // Handle *source* replica
+        let src_replica_set = shard_holder.get_shard(&transfer.shard_id);
 
-        // Should happen on a third-party side
-        // Change direction of the remote shards or add a new remote shard
-        if self.this_peer_id != transfer.from {
-            // Normally we promote the shard to become active, in case of resharding we do not.
-            // For resharding we have multiple transfers in sequence, during which the shard should
-            // remain in the resharding state. Once all are done, the shard is manually promoted to active.
-            let state = if transfer
-                .method
-                .map_or(false, |method| method.is_resharding())
-            {
-                ReplicaState::Resharding
+        if let Some(replica_set) = src_replica_set {
+            if transfer.sync || is_resharding_transfer {
+                // If transfer is *sync* (or *resharding*), we *keep* source replica
+
+                if transfer.from == self.this_peer_id {
+                    // If current peer is *transfer-sender*, we need to unproxify local shard
+
+                    replica_set.un_proxify_local().await?;
+                }
             } else {
-                ReplicaState::Active
-            };
-            let remote_shard_rerouted = transfer::driver::change_remote_shard_route(
-                shard_holder,
-                transfer.shard_id,
-                transfer.to_shard_id.unwrap_or(transfer.shard_id),
-                transfer.from,
-                transfer.to,
-                state,
-                transfer.sync,
-            )
-            .await?;
-            log::debug!("remote_shard_rerouted: {remote_shard_rerouted}");
+                // If transfer is *not* sync (and *not* resharding), we *remove* source replica
+
+                if transfer.from == self.this_peer_id {
+                    replica_set.remove_local().await?;
+                } else {
+                    replica_set.remove_remote(transfer.from).await?;
+                }
+            }
         }
-        let finish_was_registered = shard_holder.register_finish_transfer(&transfer.key())?;
-        log::debug!("finish_was_registered: {finish_was_registered}");
+
+        // Handle *destination* replica
+        let dest_replica_set =
+            shard_holder.get_shard(&transfer.to_shard_id.unwrap_or(transfer.shard_id));
+
+        if let Some(replica_set) = dest_replica_set {
+            if replica_set.peer_state(&transfer.to).is_some() && !is_resharding_transfer {
+                // Promote *destination* replica/shard to `Active` if:
```

Comment:
shouldn't we also do

```
    if to_replica_set.this_peer_id() != new_peer_id {
        to_replica_set.add_remote(new_peer_id, state).await?;
    }
```

here?

Reply:
I've checked `ShardTransfer::Start`/`ShardTransfer::Restart` and `Resharding::Start`, and I don't think we need this. We create shards (local on receiver, and remote on all other nodes) when handling `(Re)Start` messages, and so by the time we receive `ShardTransfer::Finish`, the peer should already exist in the replica set.

Reply:
Maybe there is/was some special case that needs `add_remote`, but the tests are still green without it.

Reply:
let's maybe do it anyway, just for a good measure. I assume there are no real action will happen if remote already exist?

Reply:
Meh. Sure.

Reply:
Last question:

Original impl effectively did:

```rust
let state = if is_resharding {
    ReplicaState::Resharding
} else {
    ReplicaState::Active
};

replica_set.add_remote(transfer.to, state).await?;
```

Do we want to recreate this exactly, it's fine to just do:

```rust
if !is_resharding {
    replica_set.add_remote(transfer.to, ReplicaState::Active).await?;
}
```

(If too deep/you don't have an answer right away, just answer "let's recreate exactly".)

Reply:
@generall 👆

Reply:
I don't have answer right away. I assume Resharding state is set by some other operations before transfer.

Reply:
I believe I put that there. It's so the new shard stays in resharding state. Otherwise the first transfer that finishes activates it, which we don't want.

Reply:
Tweaked a bit differently, we do `set_state`/`add_remote` always now, but at least it's still readable.

I still think this is worse than initial refactoring, though. I'd prefer this to only do absolutely minimal required changes, e.g., only _set_ (not add) replicas to `Active` in case of non-resharding transfer, and _nothing_ more. Less change of re-introducing removed replicas in the future, if/when we tweak some pre-condition checks or something.